File: python/train.py
from params import *
from MCTS import *
import time
import datetime

from RandomAgent import RandomAgent as RandomAg
from HeuristicAgent import HeuristicAgent as HeuristicAg
from MCTSAgent import MCTSAgent as MCTSAg




#ignore original network in MCTS.py and overwrite it with trainable network
network = Network(board_size = BS, input_frame_num = INPUT_FRAME_NUM, residual_num = RESIDUAL_NUM, is_trainable=True)

total_input_list = []
total_pi_list = []
total_z_list = []

for iteration in range(SELF_PLAY_NUM):
    if iteration % 10 == 0 and iteration > 0:
        #"""
        x_count = 0
        o_count = 0
        draw_count = 0
        AgentA = MCTSAg()
        AgentB = RandomAg()

        for i in range(GAME_NUM):
            if i % GAME_PRINT == 0 and i!=0:
                print(i, "  ", x_count, o_count, draw_count, "     ",x_count/i, o_count/i, draw_count/i)

            gomoku = Gomoku(BOARD_SIZE, N_IN_A_ROW)
            turn = 0
            board_state = gomoku.board

            while not gomoku.is_game_ended() and board_state.turn != BOARD_SIZE*BOARD_SIZE:
                if board_state.turn % 2 == 0:
                    board_state = AgentA.play(board_state)
                else:
                    board_state = AgentB.play(board_state)
                gomoku.board = board_state
                if gomoku.is_game_ended():
                    if board_state.turn%2*2-1 == 1:
                        x_count += 1
                    else:
                        o_count += 1
                    break
            else:
                 draw_count += 1


        print(x_count, o_count, draw_count)

        x_count = 0
        o_count = 0
        draw_count = 0
        AgentB = MCTSAg()
        AgentA = RandomAg()

        for i in range(GAME_NUM):
            if i % GAME_PRINT == 0 and i!=0:
                print(i, "  ", x_count, o_count, draw_count, "     ",x_count/i, o_count/i, draw_count/i)

            gomoku = Gomoku(BOARD_SIZE, N_IN_A_ROW)
            turn = 0
            board_state = gomoku.board

            while not gomoku.is_game_ended() and board_state.turn != BOARD_SIZE*BOARD_SIZE:
                if board_state.turn % 2 == 0:
                    board_state = AgentA.play(board_state)
                else:
                    board_state = AgentB.play(board_state)
                gomoku.board = board_state
                if gomoku.is_game_ended():
                    if board_state.turn%2*2-1 == 1:
                        x_count += 1
                    else:
                        o_count += 1
                    break
            else:
                 draw_count += 1


        print(x_count, o_count, draw_count)



        #""" 





    print("======================= iter: " + str(iteration) + " ===========================")
    
    input_list = []
    pi_list = []
    z_list = []
    
    if iteration == 0:
        game_iter = INIT_GAME_NUM
    else:
        game_iter = EPOCH_GAME_NUM

    x_count = 0
    o_count = 0
    draw_count = 0
    for game in range(game_iter):

        mcts = MCTS(BS, evaluate_with_network)
        node = mcts.root

        turn_cnt = 0
        while True:
            turn_cnt += 1
            if turn_cnt < TEMPER_THRES:
                TEMPERATURE = 1
            else:
                TEMPERATURE = TEMPER_EPS

            for i in range(MCTS_SIM_NUM):
                mcts.search(node)
            orig_state = node.state
            orig_board = orig_state.board
            input_board = preproc_board(orig_board, orig_state.turn)
            input_list.append(input_board[0])
            
            pi = np.zeros([BS * BS])
            n_list, n_sum = node.get_pi(TEMPERATURE)
            for n in n_list:
                idx = n.state.last_row * BS + n.state.last_col
                pi[idx] = n_list[n]
            pi_list.append(pi)
            
            node = mcts.play(node, TEMPERATURE)
            
            next_state = node.state
            next_board = next_state.board 
            #r = next_state.last_row
            #c = next_state.last_col

            
            if is_game_ended(next_board):
                input_board = preproc_board(next_board, next_state.turn)
                input_list.append(input_board[0])
                pi = np.zeros([BS * BS])
                n_list, n_sum = node.get_pi(TEMPERATURE)
                for n in n_list:
                    idx = n.state.last_row * BS + n.state.last_col
                    pi[idx] = n_list[n]
                pi_list.append(pi)
                nst = int(next_state.turn + 1)
                for i in range(nst):
                    if i%2*2-1 == next_state.turn%2*2-1:
                        z_list.append([1])
                    else:
                        z_list.append([-1])
            
                if next_state.turn%2*2-1 == 1:
                    winner = 'x'
                    x_count += 1
                else:
                    winner = 'o'
                    o_count += 1
                # print(next_state)
                print("game: ",game," data: ",len(input_list), \
                        '| winner:', winner, '| x: ', x_count, '| o: ', o_count, '| draw: ', draw_count, '|', \
                        x_count / (game+1), o_count / (game+1), draw_count / (game+1),
                        datetime.datetime.fromtimestamp(time.time()).strftime('%Y-%m-%d %H:%M:%S'))
                break
            elif next_state.turn == BS*BS:
                input_board = preproc_board(next_board, next_state.turn)
                input_list.append(input_board[0])
                pi = np.zeros([BS * BS])
                n_list, n_sum = node.get_pi(TEMPERATURE)
                for n in n_list:
                    idx = n.state.last_row * BS + n.state.last_col
                    pi[idx] = n_list[n]
                pi_list.append(pi)
            
                nst = next_state.turn + 1
                for _ in range(nst):
                    z_list.append([0])
            
                winner = 'draw'
                draw_count += 1
                #print(next_state)
                print("game: ",game," data: ",len(input_list), \
                        '| winner:', winner, '| x: ', x_count, '| o: ', o_count, '| draw: ', draw_count, '|', \
                        x_count / (game+1), o_count / (game+1), draw_count / (game+1),
                        datetime.datetime.fromtimestamp(time.time()).strftime('%Y-%m-%d %H:%M:%S'))
                break

    if iteration == 0:
        total_input_list = input_list
        total_pi_list = pi_list
        total_z_list = z_list
    else:
        total_input_list = input_list + total_input_list[len(input_list):]
        total_pi_list = pi_list + total_pi_list[len(pi_list):]
        total_z_list = z_list + total_z_list[len(z_list):]

    
    combined = list(zip(total_input_list, total_pi_list, total_z_list))
    random.shuffle(combined)
    total_input_list[:], total_pi_list[:], total_z_list[:] = zip(*combined)

    input_ = np.asarray(total_input_list[:BATCH_SIZE*len(input_list)])
    pi_ = np.asarray(total_pi_list[:BATCH_SIZE*len(pi_list)])
    z_ = np.asarray(total_z_list[:BATCH_SIZE*len(z_list)])
    
    
    network.train(input_, pi_, z_, EPOCH, min(len(total_input_list), BATCH_SIZE*len(input_list)), iteration)

    

File: cpp/train.cpp
#include "train.h"

#include "params.h"
#include "gomoku.h"
#include "random_agent.h"
#include "mcts_agent.h"
#include "heuristic_agent.h"
#include "network.h"

#include <stdlib.h>
#include <time.h>
#include <ctime>
#include <vector>
#include <algorithm>

using namespace std;


extern Network network;

void train()
{
	srand((unsigned)time(NULL));

	network = Network(BOARD_SIZE, 5, 9, true);

	
	for(int iteration=0; iteration<SELF_PLAY_NUM; iteration++) {
		printf("========== iter: %d ==========\n",iteration);

		vector<vector<Board> > input_list;
		vector<vector<vector<double> > > pi_list;
		vector<double> z_list;

		int game = 0;
		while((int)input_list.size() < 2*BATCH_SIZE) {
			time_t t = time(0);
			tm* now = localtime(&t);
			if(game % GAME_PRINT == 0) {
				printf("game:  %d  data:  %d  %d-%d-%d %d:%d:%d\n",game,(int)input_list.size(),\
						now->tm_year + 1900,now->tm_mon + 1,now->tm_mday,now->tm_hour,now->tm_min,now->tm_sec);
			}
			game++;

			MCTSAgent mcts_agent(MCTS_SIM_NUM);
			BoardState board_state(BOARD_SIZE, 0);

			int turn_cnt = 0;
			while(1) {
				turn_cnt++;
				
				input_list.push_back(preproc_board(board_state.board, board_state.turn));
				auto ret = mcts_agent.get_pi_and_play(board_state);
				pi_list.push_back(ret.first);
				board_state = ret.second;

				if(is_game_ended(board_state.board)) {
					input_list.push_back(preproc_board(board_state.board, board_state.turn));
					pi_list.push_back(mcts_agent.get_pi());
					int nst = board_state.turn + 1;
					for(int i=0; i<nst; i++) {
						if(i%2*2-1 == board_state.turn%2*2-1) {
							z_list.push_back(1);
						} else {
							z_list.push_back(-1);
						}
					}
					break;
				} else if(board_state.turn == BOARD_SIZE * BOARD_SIZE) {
					input_list.push_back(preproc_board(board_state.board, board_state.turn));
					pi_list.push_back(mcts_agent.get_pi());
					int nst = board_state.turn + 1;
					for(int i=0; i<nst; i++) {
						z_list.push_back(0);
					}
					break;
				}

			}
		}
		
		vector<int> index;	
		int len = (int)input_list.size();
		for(int i=0; i<len; i++){
			index.push_back(i);
		}

		for(int i=0; i<len*10; i++) {
			int s = rand() % len;
			int e = rand() % len;
			int temp = index[s];
			index[s] = index[e];
			index[e] = temp;
		}
		printf("data size: %d\n",(int)index.size());
		
		vector<vector<Board> > input_;
		vector<vector<vector<double> > > pi_;
		vector<double> z_;

		for(int i=0; i<BATCH_SIZE; i++) {
			int idx = index[i];
			input_.push_back(input_list[idx]);
			pi_.push_back(pi_list[idx]);
			z_.push_back(z_list[idx]);
		}

		network.train(input_, pi_, z_, TRAIN_ITER, PRINT_ITER);
		network.wait_training();


		int x_count = 0;
		int o_count = 0;
		int draw_count = 0;

		for(int i=0; i<GAME_NUM; i++) {
			if(i % GAME_PRINT == 0) {
				printf("%d  %d %d %d\n",i,x_count,o_count,draw_count);
			}
			RandomAgent AgentA;
			MCTSAgent AgentB;

			Gomoku gomoku(BOARD_SIZE, N_IN_A_ROW);
			int turn = 0;
			bool draw_flag = true;
			BoardState board_state = gomoku.board;

			while(!gomoku.is_game_ended() && board_state.turn != BOARD_SIZE*BOARD_SIZE) {
				if(board_state.turn % 2 == 0) {
					board_state = AgentA.play(board_state);
				} else {
					board_state = AgentB.play(board_state);
				}
				gomoku.board = board_state;

				if(gomoku.is_game_ended()) {
					if(board_state.turn%2*2-1 == 1) {
						x_count++;
					} else {
						o_count++;
					}
					draw_flag = false;
					break;
				}
			}
			if(draw_flag) {
				draw_count++;
			}

			AgentA.clear();
			AgentB.clear();
		}
		printf("win rate: %d(random) %d(mcts) %d\n\n",x_count, o_count, draw_count);

		x_count = 0;
		o_count = 0;
		draw_count = 0;

		for(int i=0; i<GAME_NUM; i++) {
			if(i % GAME_PRINT == 0) {
				printf("%d  %d %d %d\n",i,x_count,o_count,draw_count);
			}
			RandomAgent AgentB;
			MCTSAgent AgentA;

			Gomoku gomoku(BOARD_SIZE, N_IN_A_ROW);
			int turn = 0;
			bool draw_flag = true;
			BoardState board_state = gomoku.board;

			while(!gomoku.is_game_ended() && board_state.turn != BOARD_SIZE*BOARD_SIZE) {
				if(board_state.turn % 2 == 0) {
					board_state = AgentA.play(board_state);
				} else {
					board_state = AgentB.play(board_state);
				}
				gomoku.board = board_state;

				if(gomoku.is_game_ended()) {
					if(board_state.turn%2*2-1 == 1) {
						x_count++;
					} else {
						o_count++;
					}
					draw_flag = false;
					break;
				}
			}
			if(draw_flag) {
				draw_count++;
			}

			AgentA.clear();
			AgentB.clear();
		}
		printf("win rate: %d(mcts) %d(random) %d\n\n",x_count, o_count, draw_count);


	}

}
